### DISCUSSION TITLE
Looking for recursive store hash in Go

### DISCUSSION THREAD

**Discussion by onny** [nix-community/go-nix#118] [Q&A] [answered]
Hey,
I have this working C code, giving me a recursive hash for some specific directory

```
#include "store-api.hh"
#include "legacy.hh"
#include "posix-source-accessor.hh"

using namespace nix;

static int main_nix_prefetch_url(int argc, char * * argv)
{
    {

        auto store = openStore();
        auto ingestionMethod = FileIngestionMethod::Recursive;
        HashAlgorithm hashAlgo = HashAlgorithm::SHA256;
        std::optional<std::string> name = "bookmarks";
        std::optional<Hash> hash;
        Path unpacked = (Path) "/home/onny/projects/nix/bookmarks";

        PosixSourceAccessor accessor;
        auto info = store->addToStoreSlow(
            *name,
            accessor, CanonPath::fromCwd(unpacked),
            ingestionMethod, hashAlgo, {});
        assert(info.ca);
        hash = info.ca->hash;

        logger->cout(printHash16or32(hash.value()));

        return 0;
    }
}

static RegisterLegacyCommand r_nix_prefetch_url("nix-prefetch-url", main_nix_prefetch_url);
```

Returns
```
06pprhlaaqdha2nmfdcf76mhh48hdr5jlv88snxji8lpflv50wr5
```

Is there some equivalent I could use in Go?

Best regards
Jonas

**Comment by flokli** [comment, marked as answer]
Hey!

I think what you're looking for is [DumpPath](https://github.com/nix-community/go-nix/blob/93cb24a868562714f1691840e94d54ef57bc0a5a/pkg/nar/dump.go#L18), writing that to the sha256 hash function and then nixbase32-encoding it.

**Reply by onny** [reply]
thank you for the answer! i looked into the code but it's somehow difficult for me to figure this out.
I was able to use dumpPath in an example code, compile and run it. But not sure how to use it with the sha256 function :(

**Reply by flokli** [reply]
Try the following:

```go
package main

import (
	"crypto/sha256"
	"fmt"

	"github.com/nix-community/go-nix/pkg/nar"
	"github.com/nix-community/go-nix/pkg/nixbase32"
)

func main() {
	h := sha256.New()
	err := nar.DumpPath(h, "your-path")
	if err != nil {
		panic(err)
	}

	fmt.Printf("recursive store hash: %v\n", nixbase32.EncodeToString(h.Sum(nil)))
}
```

```
❯ ./foo
recursive store hash: 0cvn4pmxibandjzsqanpxdmp7z5czdzlqwdq8ddql8sm1n4kfqmr
```

`DumpPath` accepts a writer, so if you only care about the digest itself, you just literally use the sha256 hasher as a writer.

In case you also want to persist the NAR somewhere, you probably want to use some MultiWriter.

**Reply by onny** [reply]
thats great and is working really nice! going to use this in `nc4nix` to generate nextcloud package sets https://github.com/helsinki-systems/nc4nix/pull/3

**Reply by flokli** [reply]
Alright, glad it works!
